rollup/ccc/logger.go

```diff
@@ -0,0 +1,661 @@
+	1,  // SWAP9 (152)
+	1,  // SWAP10 (153)
+	1,  // SWAP11 (154)
+	1,  // SWAP12 (155)
+	1,  // SWAP13 (156)
+	1,  // SWAP14 (157)
+	1,  // SWAP15 (158)
+	1,  // SWAP16 (159)
+	2,  // LOG0 (160)
+	2,  // LOG1 (161)
+	2,  // LOG2 (162)
+	2,  // LOG3 (163)
+	2,  // LOG4 (164)
+	0,  // UNDEFINED (165)
+	0,  // UNDEFINED (166)
+	0,  // UNDEFINED (167)
+	0,  // UNDEFINED (168)
+	0,  // UNDEFINED (169)
+	0,  // UNDEFINED (170)
+	0,  // UNDEFINED (171)
+	0,  // UNDEFINED (172)
+	0,  // UNDEFINED (173)
+	0,  // UNDEFINED (174)
+	0,  // UNDEFINED (175)
+	0,  // UNDEFINED (176)
+	0,  // UNDEFINED (177)
+	0,  // UNDEFINED (178)
+	0,  // UNDEFINED (179)
+	0,  // UNDEFINED (180)
+	0,  // UNDEFINED (181)
+	0,  // UNDEFINED (182)
+	0,  // UNDEFINED (183)
+	0,  // UNDEFINED (184)
+	0,  // UNDEFINED (185)
+	0,  // UNDEFINED (186)
+	0,  // UNDEFINED (187)
+	0,  // UNDEFINED (188)
+	0,  // UNDEFINED (189)
+	0,  // UNDEFINED (190)
+	0,  // UNDEFINED (191)
+	0,  // UNDEFINED (192)
+	0,  // UNDEFINED (193)
+	0,  // UNDEFINED (194)
+	0,  // UNDEFINED (195)
+	0,  // UNDEFINED (196)
+	0,  // UNDEFINED (197)
+	0,  // UNDEFINED (198)
+	0,  // UNDEFINED (199)
+	0,  // UNDEFINED (200)
+	0,  // UNDEFINED (201)
+	0,  // UNDEFINED (202)
+	0,  // UNDEFINED (203)
+	0,  // UNDEFINED (204)
+	0,  // UNDEFINED (205)
+	0,  // UNDEFINED (206)
+	0,  // UNDEFINED (207)
+	0,  // UNDEFINED (208)
+	0,  // UNDEFINED (209)
+	0,  // UNDEFINED (210)
+	0,  // UNDEFINED (211)
+	0,  // UNDEFINED (212)
+	0,  // UNDEFINED (213)
+	0,  // UNDEFINED (214)
+	0,  // UNDEFINED (215)
+	0,  // UNDEFINED (216)
+	0,  // UNDEFINED (217)
+	0,  // UNDEFINED (218)
+	0,  // UNDEFINED (219)
+	0,  // UNDEFINED (220)
+	0,  // UNDEFINED (221)
+	0,  // UNDEFINED (222)
+	0,  // UNDEFINED (223)
+	0,  // UNDEFINED (224)
+	0,  // UNDEFINED (225)
+	0,  // UNDEFINED (226)
+	0,  // UNDEFINED (227)
+	0,  // UNDEFINED (228)
+	0,  // UNDEFINED (229)
+	0,  // UNDEFINED (230)
+	0,  // UNDEFINED (231)
+	0,  // UNDEFINED (232)
+	0,  // UNDEFINED (233)
+	0,  // UNDEFINED (234)
+	0,  // UNDEFINED (235)
+	0,  // UNDEFINED (236)
+	0,  // UNDEFINED (237)
+	0,  // UNDEFINED (238)
+	0,  // UNDEFINED (239)
+	9,  // CREATE (240)
+	12, // CALL (241)
+	12, // CALLCODE (242)
+	4,  // RETURN (243)
+	12, // DELEGATECALL (244)
+	9,  // CREATE2 (245)
+	0,  // UNDEFINED (246)
+	0,  // UNDEFINED (247)
+	0,  // UNDEFINED (248)
+	0,  // UNDEFINED (249)
+	12, // STATICCALL (250)
+	0,  // UNDEFINED (251)
+	0,  // UNDEFINED (252)
+	4,  // REVERT (253)
+	0,  // INVALID (254)
+	0,  // SELFDESTRUCT (255)
+}
+
+func constantStateUsage(usage uint64) func(*vm.ScopeContext, int) uint64 {
+	return func(_ *vm.ScopeContext, _ int) uint64 {
+		return usage
+	}
+}
+
+func logStateUsage(size uint64) func(*vm.ScopeContext, int) uint64 {
+	return func(scope *vm.ScopeContext, _ int) uint64 {
+		return 2*(scope.Stack.Back(1).Uint64()/32) + 7 + 2*size
+	}
+}
+
+// state circuit resource usage per OpCode
+var stateUsagePerOpCode = [256]func(*vm.ScopeContext, int) uint64{
```

Comment:
Should we use mapping instead of array, like here https://github.com/scroll-tech/go-ethereum/blob/develop/core/vm/jump_table.go#L245? To avoid some obvious but painful bugs.

Reply:
That weirdly looks like a map syntax but JumpTable is actually an array. https://github.com/scroll-tech/go-ethereum/blob/develop/core/vm/jump_table.go#L67